### PR TITLE
Wire replication stream E2E

### PR DESCRIPTION
**Pull request by wxing1292**
**What changed?**
* Add replication stream API &UT

**Why?**
Support replication stream

**How did you test it?**
UT & integration tests

**Potential risks**
N/A

**Is hotfix candidate?**
N/A

### CODE REVIEW

**Comment by wxing1292** [comment]
review #4093 first

**Comment by wxing1292** [comment]
review #4107 first

File: Makefile [hunk 1]
```diff
@@ -116,7 +116,7 @@ update-linters:
 
 update-mockgen:
 	@printf $(COLOR) "Install/update mockgen tool..."
-	@go install github.com/golang/mock/mockgen@v1.6.0
+	@go install github.com/golang/mock/mockgen@v1.7.0-rc.1
```
Comment (yux0) [inline Makefile:119]: any feature we need in this rc version?

File: common/dynamicconfig/constants.go [hunk 2]
```diff
@@ -338,6 +338,9 @@ const (
 
 	// keys for history
 
+	// EnableReplicationStream turn on replication stream
+	EnableReplicationStream = "history,enableReplicationStream"
```
Comment (yux0) [inline common/dynamicconfig/constants.go:342]: nit: history.enableReplicationStream

File: service/history/handler.go [hunk 3]
```diff
@@ -63,9 +63,9 @@ import (
 	"go.temporal.io/server/common/searchattribute"
 	serviceerrors "go.temporal.io/server/common/serviceerror"
 	"go.temporal.io/server/service/history/api"
+	replication2 "go.temporal.io/server/service/history/api/replication"
```
Comment (yux0) [inline service/history/handler.go:66]: nit:replicaitonclient?

File: service/history/replication/task_processor_manager.go [hunk 4]
```diff
@@ -131,7 +134,9 @@ func (r *taskProcessorManagerImpl) Start() {
 	}
 
 	// Listen to cluster metadata and dynamically update replication processor for remote clusters.
-	r.listenToClusterMetadataChange()
+	if r.enableFetcher {
```
Comment (yux0) [inline service/history/replication/task_processor_manager.go:137]: with streaming disable, it should listen to metadata change as well?

File: service/history/replication/ack_manager.go [hunk 5]
```diff
@@ -503,3 +471,99 @@ func (p *ackMgrImpl) clusterToReaderID(
...
+}
+
+func (p *ackMgrImpl) SubscribeNotification() (<-chan struct{}, string) {
+	subscriberID := uuid.New().String()
+
+	p.subscriberLock.Lock()
+	defer p.subscriberLock.Unlock()
+
+	for {
+		if _, ok := p.subscribers[subscriberID]; !ok {
```
Comment (meiliang86) [inline service/history/replication/ack_manager.go:515]: How can this be not true? uuid collision?

File: service/history/api/replication/stream.go [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+	return catchupEndExclusiveWatermark.TaskID, nil
+}
+
+func sendLive(
+	ctx context.Context,
+	server historyservice.HistoryService_StreamWorkflowReplicationMessagesServer,
+	shardContext shard.Context,
+	sourceClusterShardID historyclient.ClusterShardID,
+	targetClusterShardID historyclient.ClusterShardID,
+	channel <-chan struct{},
```
Comment (meiliang86) [inline service/history/api/replication/stream.go:203]: nit: rename to newTaskNotificationChan

File: service/history/replication/stream_receiver_monitor.go [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+		status       int32
+		shutdownOnce channel.ShutdownOnce
+
+		sync.Mutex
+		streams map[ClusterShardKeyPair]*StreamReceiver
+	}
+)
+
+func NewStreamReceiverMonitor(
```
Comment (meiliang86) [inline service/history/replication/stream_receiver_monitor.go:52]: Add some unit test for this file?

**Review by meiliang86 — APPROVED** [review]

File: Makefile (hunk 1 above)
Comment (wxing1292) [reply Makefile:119]: the ability for mockgen to handle generics: https://github.com/temporalio/temporal/pull/4097/files#diff-1f9d74f27f59376896675188a3673cac1730b7afc79e890ca49f0da83b3f23e6R105

File: service/history/shard/engine.go [hunk 8]
```diff
@@ -90,5 +91,22 @@ type (
 
 		NotifyNewHistoryEvent(event *events.Notification)
 		NotifyNewTasks(tasks map[tasks.Category][]tasks.Task)
+
+		ReplicationStream
+	}
+
+	ReplicationStream interface {
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/shard/engine.go:98]: i can give it a try, but probably circular imports

File: service/history/replication/task_processor_manager.go [hunk 9]
```diff
@@ -70,6 +70,7 @@ type (
 		metricsHandler                metrics.Handler
 		logger                        log.Logger
 
+		enableFetcher     bool
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/replication/task_processor_manager.go:73]: this is actually correct, `enableFetcher:        !config.EnableReplicationStream(),`

File: service/history/replication/stream_receiver_monitor.go (hunk 7 above)
Comment (wxing1292) [reply service/history/replication/stream_receiver_monitor.go:52]: UT for this logic will be added later

File: service/history/replication/task_processor_manager.go (hunk 4 above)
Comment (wxing1292) [reply service/history/replication/task_processor_manager.go:137]: this is fetcher specific
there is also a stream specific var: https://github.com/temporalio/temporal/pull/4097/files#diff-9c489cc6ae52b9289979f47fcafaca2691ad9e7bf4e537597a4bbf68913b0217R42

File: service/history/replication/ack_manager.go (hunk 5 above)
Comment (wxing1292) [reply service/history/replication/ack_manager.go:515]: not ture == no collision => common case
true == uuid collision, retry

**Pull request merged**